packages/chain-adapters/src/evm/EvmBaseAdapter.ts

```diff
@@ -276,6 +290,27 @@ export abstract class EvmBaseAdapter<T extends EvmChainId> implements IChainAdap
   async signAndBroadcastTransaction(signTxInput: SignTxInput<ETHSignTx>): Promise<string> {
     try {
       const { txToSign, wallet } = signTxInput
+      const assets = await getGeneratedAssetData()
+      const feeAsset = assets[this.getFeeAssetId()]
+      // If there is a mismatch between the current wallet's EVM chain ID and the adapter's chainId?
+      // Switch the chain on wallet before building/sending the Tx
+      if (supportsEthSwitchChain(wallet)) {
+        const walletEvmChainId = await (wallet as ETHWallet).ethGetChainId?.()
+        const adapterEvmChainId = fromChainId(this.chainId).chainReference
+        if (!bnOrZero(walletEvmChainId).isEqualTo(Number(adapterEvmChainId))) {
```

Comment:
There is a potential bug here. We can contrive a case where `walletEvmChainId` is invalid (which makes it `0`), and the chain adapters `ChainReference` is `0` (I don't think this exists in the real world, but it's possible). 

In this case the expression would evaluate to `false` and the logic inside the if statement would not be hit. Edge case, but worth noting.

Reply:
Potential security concern actually, TIL, see this draft EIP https://eips.ethereum.org/EIPS/eip-3788
81bbfa3625dddd5d94b491b89bd7e383c11e036b